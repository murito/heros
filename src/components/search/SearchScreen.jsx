import React, { useMemo } from 'react'
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heros/HeroCard';
import { searchHeros } from '../../selectors/filters';

import './SearchScreen.scss';

export const SearchScreen = ({ history }) => {
    // get location and parse query string
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    
    // useForm manage form changes
    const [ { search }, handleInputChange ] = useForm({ search: q});

    // Filter data
    const herosFiltered = useMemo(() => searchHeros(q), [ q ]);

    // Form submit management
    const handleSearch = (e) => {
        e.preventDefault();

        history.push(`?q=${ search }`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-md-4">
                    <h4>Search Form</h4>
                    <form onSubmit={ handleSearch }>
                        <div className="form-group">
                            <input type="text" 
                                placeholder="Find your hero here..."
                                className="form-control"
                                name="search"
                                autoComplete="off"
                                value={ search }
                                onChange={ handleInputChange }
                            />
                        </div>
                        
                        <button type="submit"
                            className="btn btn-block btn-outline-primary">
                                Search
                        </button>
                    </form>
                </div>

                <div className="col-md-8">
                    <h4>{ herosFiltered.length } results for "{ q }"</h4>
                    <hr/>
                    <div className="results">
                        { q === '' &&  <small className="alert alert-info">No results here!</small> }
                        {
                            herosFiltered.map( hero => 
                                <HeroCard {...hero} key={ hero.id } />   
                            )
                        }
                    </div>
                </div>
            </div>
        </div>
    )
}
