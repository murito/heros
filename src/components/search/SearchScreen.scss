.results{
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

.alert{
    width: 100%;
    text-align: center;
}