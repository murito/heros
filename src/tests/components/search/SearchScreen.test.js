import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Tests on <SearchScreen />', () => {
    const historyMock = {
        push: jest.fn()
    };

    const wrapper = mount(
        <MemoryRouter initialEntries={ ['/search'] }>
            <Route path="/search" component={ SearchScreen }/>
        </MemoryRouter>
    );

    test('should be shown itself with the default values', () => {    
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe( 'No results here!' );
    });
    
    test('should show Batman and the input with the queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );

        expect( wrapper.find('h4').at(1).text().trim() ).toBe(`1 results for "batman"`);
        expect( wrapper ).toMatchSnapshot();
    })
    
    test('should call the push on the history object', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <Route 
                    path="/search" 
                    component={ () => <SearchScreen history={ historyMock }/> }/>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'batman'
            }
        });
        
        wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

        expect( historyMock.push ).toHaveBeenCalledWith('?q=batman');
    })
    
})
