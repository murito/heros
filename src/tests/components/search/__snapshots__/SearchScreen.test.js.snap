// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchScreen /> should be shown itself with the default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="search"
            onChange={[Function]}
            placeholder="Find your hero here..."
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-md-8"
    >
      <h4>
        0
         results for "
        "
      </h4>
      <hr />
      <div
        className="results"
      >
        <small
          className="alert alert-info"
        >
          No results here!
        </small>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests on <SearchScreen /> should show Batman and the input with the queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="search"
            onChange={[Function]}
            placeholder="Find your hero here..."
            type="text"
            value="batman"
          />
        </div>
        <button
          className="btn btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-md-8"
    >
      <h4>
        1
         results for "
        batman
        "
      </h4>
      <hr />
      <div
        className="results"
      >
        <div
          className="card ms-3 hero-card"
        >
          <img
            alt="Batman"
            className="card-img"
            src=""
          />
          <div
            className="card-body"
          >
            <div
              className="characters dc"
            >
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                <h4
                  className="text-center"
                >
                  Bruce Wayne
                </h4>
              </a>
            </div>
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <small
              className="text-muted"
            >
              Detective Comics #27
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
